cmake/D2D_MPI.cmake

```diff
@@ -0,0 +1,51 @@
+# MPI CMakeLists
+
+find_package(MPI REQUIRED)
+# Stop if there is no MPI_Fortran_Compiler
+if (MPI_Fortran_COMPILER)
+    message(STATUS "MPI_Fortran_COMPILER found: ${MPI_Fortran_COMPILER}")
+    message(STATUS "MPI_VERSION found: ${MPI_VERSION}")
+    # Try to guess the MPI type to adapt compilation flags if necessary
+    string(FIND "${MPI_Fortran_COMPILER}" "mpich" pos)
+    if(pos GREATER_EQUAL "0")
+      set(FIND_MPICH TRUE)
+      message(STATUS "MPI is MPICH type")
+    endif()
+    string(FIND "${MPI_Fortran_COMPILER}" "openmpi" pos)
+    if(pos GREATER_EQUAL "0")
+      set(FIND_OMPI TRUE)
+      message(STATUS "MPI is openMPI type")
+    endif()
+else (MPI_Fortran_COMPILER)
+    message(SEND_ERROR "This application cannot compile without MPI")
+endif(MPI_Fortran_COMPILER)
+# Warning if Include are not found => can be fixed with more recent cmake version
+if (MPI_FOUND)
+    message(STATUS "MPI FOUND: ${MPI_FOUND}")
+    include_directories(SYSTEM ${MPI_INCLUDE_PATH})
+    message(STATUS "MPI INCL ALSO FOUND: ${MPI_INCLUDE_PATH}")
+    if (NOT MPI_NUMPROCS_SET)
+      message(STATUS "Reset the number of ranks to 1")
+      set(MPIEXEC_MAX_NUMPROCS "1" CACHE STRING
+          "Set the initial value to 1 rank" FORCE)
+      set(MPI_NUMPROCS_SET 1 CACHE INTERNAL "MPI Ranks set")
+      # Force the mpirun to be coherent with the mpifortran
+      string(REPLACE "mpif90" "mpirun" PATH_TO_MPIRUN "${MPI_Fortran_COMPILER}")
```

Comment:
When does the mpiexec found by FindMPI not work? In my experience as long as I pass FC to cmake it finds the right one. This solution does not work with `FC=mpiiftort` or if trying to specify a particular mpif90, e.g. `FC=mpif90.openmpi`

Reply:
A partial solution in the case of `mpiifort` is to do the string replace twice:
```
string(REPLACE "mpif90" "mpirun" PATH_TO_MPIRUN "${MPI_Fortran_COMPILER}")
string(REPLACE "mpiifort" "mpirun" PATH_TO_MPIRUN "${MPI_Fortran_COMPILER}")
```
but that doesn't resolve the issue for `mpif90.openmpi` - the same solution should work, but this could quickly get out of hand

Reply:
The issue is not `mpif90` but mpiexec/mpirun. Most of the time `MPIEXEC_EXECUTABLE` points to a random MPI (not really random but the first in alphabetical order) or it is not found (for example in my Mac). This is not particularly an issue until running the test. For example in your case with NVHPC that the test is not running i think it is because it is pointing to the wrong mpirun, despite all the rest being correct and coherent.

Reply:
In my case for the Mac it is ok since I have `mpif90-openmpi-mp` and the equivalent `mpirun-openmpi-mp`

Reply:
It fails on `mpif90.openmpi` for me

Reply:
Durr - need to replace the variable being searched in the second REPLACE operation